feat(email): Add cooldown period to prevent rapid resend of verificat…

smoshie25 opened the pull request:
### Summary

This PR introduces a realm-configurable cooldown period between verification email resends to prevent users from triggering excessive emails by repeatedly clicking the "Resend Email" link.

Fixes [#41331](https://github.com/keycloak/keycloak/issues/41331)

### Details

- Introduces support for a realm-level attribute: `emailResendCooldownSeconds`
- If not defined, falls back to a default cooldown of 30 seconds
- Stores `emailLastSent` timestamp in the user's authentication session
- Respects the cooldown on each resend attempt

### Configuration

To set the cooldown:

```bash
curl -X PUT http://localhost:8080/admin/realms/<realm> \
  -H "Authorization: Bearer <token>" \
  -H "Content-Type: application/json" \
  -d '{
    "attributes": {
      "emailResendCooldownSeconds": "60"
    }
  }'

ahus1 reviewed (dismissed):
Thank you for looking into this task. 

Some time last year, the required actions became configurable via the admin UI. The UpdatePassword required action was the first - see ab376d91012ce7168256a2f198287eafed83ea35 for the change. This would allow users to configure it by realm and via the UI and the API.  

As an alternative, there is a the SingleUseObjectProvider where you can put items into a cache that is shared across nodes and authentication sessions. See https://github.com/keycloak/keycloak/blob/478e0b3264e59fad467ad4c4a2b0237487c81749/services/src/main/java/org/keycloak/authentication/authenticators/client/JWTClientValidator.java#L250-L253 for an example. 

It would be good if the error message could be a bit more descriptive by showing the number of seconds it would take for the next resend to be allowed. 

A formal requirement for all PRs: They need a sign-off as described in the CONTRIBUTING.md file. 

Eventually we would also like to see a test - I can help with that once the functionality is stable.

Commit:
feat(requiredaction): add configurable cooldown for email resend in VerifyEmail

Introduces support for configuring the cooldown period between email resends in the VerifyEmail required action.

This enhancement allows administrators to set the cooldown duration via the admin console, improving flexibility and aligning UI behavior with actual enforcement logic.

Signed-off-by: Moshie Samuel <moshie.samuel@gmail.com>
Signed-off-by: moshiem <moshiem@hardcorebiometric.com>

ahus1 committed:
Rework

Signed-off-by: Alexander Schwartz <aschwart@redhat.com>

ahus1 committed:
Adding docs

Signed-off-by: Alexander Schwartz <aschwart@redhat.com>

ahus1 committed:
Refactoring

Signed-off-by: Alexander Schwartz <aschwart@redhat.com>

ahus1 reviewed (commented):
I've reviewed the change, and pushed an additional commit. 

Here's what I changed on a high level:

* When the user is presented with the verify email required action, an email is sent automatically. To avoid showing the user an error message for this initial email, the email will always be sent out for this flow. 
* The message is now localizable
* I found a way to store the expiry in the map, so we can show it in the error message
* There is now a test

Something I considered but then did not do: 

* Setting the entry in the cache only after the email is sent out successfully. I see that it will redirect to an error page anyway, and wouldn't keep the user on the same page. 

Please review and let me know your thoughts. I'll be triggering the CI, let's see what that brings.

ahus1 committed:
Fixing one more test

Signed-off-by: Alexander Schwartz <aschwart@redhat.com>

ahus1 committed:
Fixing tests

Signed-off-by: Alexander Schwartz <aschwart@redhat.com>
